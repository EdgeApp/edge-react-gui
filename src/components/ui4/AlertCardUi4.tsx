import * as React from 'react'
import { View } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { CardUi4 } from './CardUi4'

interface Props {
  body?: string[] | string // Bullet point messages if an array is provided
  title: string
  type: 'error' | 'warning'
  footer?: string
  header?: string

  // DO NOT USE after a scene is fully UI4! Margins should all align without adjustment.
  marginRem?: number[] | number

  onPress?: () => void
}
/**
 * A warning or error card that accepts a title, header, bullet points OR normal
 * message, and a footer. The body is considered bullet pointed if an array of
 * strings is given.
 * Bullet points, if provided, are rendered with the correct vertical alignment when
 * the messages overflow.
 *  .___________________________.
 *  | ⚠ Title Text              |
 *  |                           |
 *  | This is the header text   |
 *  |                           |
 *  | • This is an overflowing  |
 *  |   bulletpoint message     |
 *  | • This one's short        |
 *  |                           |
 *  | This is the footer text   |
 *  |___________________________|
 */
export function AlertCardUi4(props: Props) {
  const { title, type, header, body, footer, marginRem, onPress } = props
  const theme = useTheme()
  const styles = getStyles(theme)

  const bulletPoint = (bulletPoint: string) => {
    return (
      <View style={styles.bulletpointRow} key={bulletPoint}>
        <EdgeText style={styles.bulletpointText}>{'\u2022 '}</EdgeText>
        <EdgeText style={styles.bulletpointText} numberOfLines={0}>
          {bulletPoint}
        </EdgeText>
      </View>
    )
  }

  return (
    <CardUi4 gradientBackground={type === 'error' ? theme.cardGradientError : theme.cardGradientWarning} marginRem={marginRem} onPress={onPress}>
      <View style={styles.container}>
        <View style={styles.titleContainer}>
          <IonIcon name="warning-outline" style={styles.icon} color={theme.primaryText} size={theme.rem(1.25)} />
          <EdgeText numberOfLines={0} style={styles.titleText}>
            {title}
          </EdgeText>
        </View>

        {header == null ? null : <EdgeText style={styles.text}>{header}</EdgeText>}

        {body == null ? null : typeof body === 'object' && body.length > 0 ? (
          <View style={styles.bulletPointContainer}>{body.map(point => bulletPoint(point))}</View>
        ) : (
          <EdgeText style={styles.text} numberOfLines={10}>
            {body}
          </EdgeText>
        )}

        {footer == null ? null : (
          <EdgeText style={styles.text} numberOfLines={10}>
            {footer}
          </EdgeText>
        )}
      </View>
    </CardUi4>
  )
}

const getStyles = (theme: Theme) =>
  cacheStyles((theme: Theme) => ({
    container: {
      margin: theme.rem(0.5)
    },
    titleContainer: {
      flexDirection: 'row',
      alignItems: 'center'
    },
    titleText: {
      marginLeft: theme.rem(0.2),
      fontFamily: theme.fontFaceMedium,
      flexShrink: 1
    },
    icon: {
      marginRight: theme.rem(0.2)
    },
    text: {
      fontSize: theme.rem(0.75),
      marginHorizontal: theme.rem(0.25),
      marginTop: theme.rem(0.5)
    },
    bulletPointContainer: {
      marginTop: theme.rem(0.5)
    },
    bulletpointRow: {
      flexDirection: 'row'
    },
    bulletpointText: {
      marginLeft: theme.rem(0.2),
      fontSize: theme.rem(0.75)
    }
  }))(theme)
