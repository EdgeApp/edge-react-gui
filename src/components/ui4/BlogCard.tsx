import { BlogPost } from 'edge-info-server/types'
import * as React from 'react'
import { View } from 'react-native'
import FastImage from 'react-native-fast-image'

import { useHandler } from '../../hooks/useHandler'
import { getLocaleOrDefaultString } from '../../locales/intl'
import { openBrowserUri } from '../../util/WebUtils'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { CardUi4 } from './CardUi4'

interface Props {
  blogPost: BlogPost
}

const IMAGE_HEIGHT_RATIO = '65%'

/**
 * Blog post card with a top image and text below.
 */
export const BlogCard = (props: Props) => {
  const { localeTitle, localeBody, localeBlogUrl, lightImageUrl, darkImageUrl } = props.blogPost

  const theme = useTheme()
  const styles = getStyles(theme)

  const title = getLocaleOrDefaultString(localeTitle)
  const body = getLocaleOrDefaultString(localeBody)
  const url = getLocaleOrDefaultString(localeBlogUrl)
  const image = theme.isDark ? darkImageUrl : lightImageUrl

  const handlePress = useHandler(() => {
    if (url != null) openBrowserUri(url)
  })

  return (
    <CardUi4
      onPress={handlePress}
      nodeBackground={
        <View style={styles.nodeBackground}>
          <FastImage source={{ uri: image }} style={styles.bannerImage} />
        </View>
      }
    >
      <View style={styles.backgroundSpacing} />
      <View style={styles.textContainer}>
        <EdgeText style={[styles.titleText, theme.cardTextShadow]} numberOfLines={1}>
          {title}
        </EdgeText>
        <EdgeText style={[styles.bodyText, theme.cardTextShadow]} numberOfLines={2} disableFontScaling>
          {body}
        </EdgeText>
      </View>
    </CardUi4>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  backgroundSpacing: {
    height: IMAGE_HEIGHT_RATIO
  },
  textContainer: {
    margin: theme.rem(0.5)
  },
  titleText: {
    fontFamily: theme.fontFaceMedium,
    marginBottom: theme.rem(0.25)
  },
  bodyText: {
    fontSize: theme.rem(0.65)
  },
  nodeBackground: {
    height: IMAGE_HEIGHT_RATIO,
    justifyConent: 'center',
    alignItems: 'center'
  },
  bannerImage: {
    height: '100%',
    width: '100%'
  }
}))
