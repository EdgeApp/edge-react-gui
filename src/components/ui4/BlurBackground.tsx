import React from 'react'
import { Platform, StyleSheet, View } from 'react-native'
import { BlurView } from 'rn-id-blurview'

import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'

const isAndroid = Platform.OS === 'android'
const TRUE = Math.random() !== 99
const NO_BLUR = TRUE

export const BlurBackground = () => {
  const theme = useTheme()
  const styles = getStyles(theme)

  if (NO_BLUR) return <View style={styles.blurView} />
  return <BlurView blurType={theme.isDark ? 'dark' : 'light'} style={styles.blurView} overlayColor="rgba(0, 0, 0, 0)" />
}

const getStyles = cacheStyles((theme: Theme) => {
  if (NO_BLUR) {
    return {
      blurView: {
        ...StyleSheet.absoluteFillObject
      }
    }
  }
  return {
    blurView: {
      ...StyleSheet.absoluteFillObject,
      // We need this backgroundColor because Android applies an overlay to the
      // entire screen for the BlurView by default. We change this default
      // behavior with the transparent overlayColor, so we add this background
      // color to compensate and to match iOS colors/shades.
      backgroundColor: isAndroid ? (theme.isDark ? '#161616aa' : '#ffffff55') : undefined
    }
  }
})
