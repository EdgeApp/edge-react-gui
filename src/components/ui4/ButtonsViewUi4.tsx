/**
 * IMPORTANT: Changes in this file MUST be synced between edge-react-gui and
 * edge-login-ui-rn!
 */

import * as React from 'react'
import { Platform, View, ViewStyle } from 'react-native'

import { EdgeAnim } from '../common/EdgeAnim'
import { maybeComponent } from '../hoc/maybeComponent'
import { styled } from '../hoc/styled'
import { Space } from '../layout/Space'
import { ButtonTypeUi4, ButtonUi4 } from './ButtonUi4'

const INTER_BUTTON_SPACING_REM = 1
const ANIM_DURATION = 1000
const ANIM_DISTANCE_INCREMENT = 20

export interface ButtonInfo {
  label: string
  onPress: () => void | Promise<void>
  disabled?: boolean
  spinner?: boolean
  testID?: string

  animDistanceStart?: number
}

interface Props {
  // Specifies whether the component should be positioned absolutely.
  // Default value is false.
  absolute?: boolean
  // ButtonInfos
  primary?: ButtonInfo
  secondary?: ButtonInfo
  secondary2?: ButtonInfo // A secondary-styled button in the primary position (right/top side)
  tertiary?: ButtonInfo

  // Arrangement of the button(s). Defaults to 'column' or 'solo' depending on
  // number of ButtonInfos given
  layout?:
    | 'row' // Buttons are stacked side by side horizontally, taking up 50% of the available space each.
    | 'column' // Buttons stacked on top of each other vertically, taking up as much space as the widest button.
    | 'solo' // A single centered button whose size is determined by label length (default for single-button props)

  // What kind of component this ButtonsView lives on. Affects margins.
  parentType?: 'scene' | 'modal'

  // 'distance' prop of the first button
  animDistanceStart?: number
}

/**
 * A consistently styled view for displaying button layouts.
 */
export const ButtonsViewUi4 = React.memo(
  ({ absolute = false, primary, secondary, secondary2, tertiary, layout = 'column', parentType, animDistanceStart }: Props) => {
    const buttonInfos = [primary, secondary, secondary2, tertiary].filter(key => key != null)
    if (buttonInfos.length === 1) layout = 'solo'

    const spacing = <Space around={INTER_BUTTON_SPACING_REM / 2} />

    const renderButton = (type: ButtonTypeUi4, buttonProps?: ButtonInfo, index: number = 0) => {
      if (buttonProps == null) return null
      const { label, onPress, disabled, spinner, testID } = buttonProps

      const distance = animDistanceStart != null ? animDistanceStart + index * ANIM_DISTANCE_INCREMENT : undefined
      // TODO: Sync EdgeAnim w/ LoginUi
      const disableAnimation = Platform.OS === 'android'

      return (
        <MaybeEdgeAnim when={animDistanceStart != null} disableAnimation={disableAnimation} enter={{ type: 'fadeInDown', duration: ANIM_DURATION, distance }}>
          <ButtonUi4 layout={layout} label={label} onPress={onPress} type={type} disabled={disabled} spinner={spinner} testID={testID} />
        </MaybeEdgeAnim>
      )
    }

    const hasPrimary = primary != null
    const hasSecondary = secondary != null
    const hasSecondary2 = secondary2 != null
    const hasTertiary = tertiary != null

    return (
      <StyledButtonContainer absolute={absolute} layout={layout} parentType={parentType}>
        {hasPrimary && (
          <>
            {renderButton('primary', primary, 0)}
            {(hasSecondary || hasSecondary2 || hasTertiary) && spacing}
          </>
        )}
        {hasSecondary && (
          <>
            {renderButton('secondary', secondary, hasPrimary ? 1 : 0)}
            {(hasSecondary2 || hasTertiary) && spacing}
          </>
        )}
        {hasSecondary2 && (
          <>
            {renderButton('secondary', secondary2, (hasPrimary ? 1 : 0) + (hasSecondary ? 1 : 0))}
            {hasTertiary && spacing}
          </>
        )}
        {hasTertiary && <>{renderButton('tertiary', tertiary, (hasPrimary ? 1 : 0) + (hasSecondary ? 1 : 0) + (hasSecondary2 ? 1 : 0))}</>}
      </StyledButtonContainer>
    )
  }
)

/** @deprecated - Shouldn't use this post-UI4 transition once all our layouts have been codified into this component. */
export const StyledButtonContainer = styled(View)<{
  absolute?: boolean
  layout: 'row' | 'column' | 'solo'
  parentType?: 'scene' | 'modal'
}>(theme => props => {
  const { absolute, layout, parentType } = props

  const marginSize = theme.rem(0.5)

  const baseStyle: ViewStyle = {
    margin: marginSize
  }

  const absoluteStyle: ViewStyle = absolute
    ? {
        position: 'absolute',
        bottom: 0,
        left: marginSize,
        right: marginSize
      }
    : {}

  const soloStyle: ViewStyle =
    layout === 'solo'
      ? {
          justifyContent: 'center',
          marginHorizontal: theme.rem(0.5),
          alignItems: 'center',
          // TODO: This causes too much space in AddressModal.
          // Move flex to sceneMarginStyle, use in AddressModal, sync w/ LoginUi
          flexGrow: 1,
          flexShrink: 1
        }
      : {}

  const rowStyle: ViewStyle =
    layout === 'row'
      ? {
          flexGrow: 1,
          flexShrink: 1,
          flexDirection: 'row-reverse',
          justifyContent: 'center'
        }
      : {}

  const columnStyle: ViewStyle =
    layout === 'column'
      ? {
          alignSelf: 'center', // Shrink view around buttons
          alignItems: 'stretch', // Stretch our children out
          flexDirection: 'column',
          justifyContent: 'space-between'
        }
      : {}

  const sceneMarginStyle: ViewStyle =
    parentType === 'scene'
      ? {
          marginBottom: theme.rem(3),
          marginTop: theme.rem(1)
        }
      : {}

  const modalMarginStyle: ViewStyle =
    parentType === 'modal'
      ? {
          marginBottom: theme.rem(1),
          marginTop: theme.rem(2)
        }
      : {}

  return {
    ...baseStyle,
    ...absoluteStyle,
    ...soloStyle,
    ...rowStyle,
    ...columnStyle,
    ...sceneMarginStyle,
    ...modalMarginStyle
  }
})

const MaybeEdgeAnim = maybeComponent(EdgeAnim)
