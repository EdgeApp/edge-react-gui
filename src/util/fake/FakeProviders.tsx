import * as React from 'react'
import { Metrics, SafeAreaProvider } from 'react-native-safe-area-context'
import { Provider } from 'react-redux'
import { applyMiddleware, createStore } from 'redux'
import thunk from 'redux-thunk'

import { rootReducer, RootState } from '../../reducers/RootReducer'

type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>
    }
  : T

export type FakeState = DeepPartial<RootState>

interface Props {
  children: React.ReactNode
  initialState?: FakeState
}

export function FakeProviders(props: Props) {
  const { children, initialState = {} } = props

  const store = React.useMemo(() => createStore(rootReducer, initialState as any, applyMiddleware(thunk)), [initialState])
  return (
    <SafeAreaProvider initialMetrics={initialMetrics}>
      <Provider store={store}>{children}</Provider>
    </SafeAreaProvider>
  )
}

// These match what our old react-native-safe-area-view library returned:
const initialMetrics: Metrics = {
  frame: { height: 1334, width: 750, x: 0, y: 0 },
  insets: { bottom: 0, left: 0, right: 0, top: 20 }
}
