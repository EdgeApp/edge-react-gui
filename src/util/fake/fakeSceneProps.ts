import { AppParamList, EdgeSceneProps, NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener() {
    return () => undefined
  },
  removeListener() {},
  isFocused() {
    return true
  },

  navigate() {},
  push() {},
  replace() {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  dispatch() {},
  reset() {},
  canGoBack() {
    return true
  },
  getId() {
    return undefined
  },
  getState() {
    return {
      key: 'foo',
      index: 0,
      routes: [],
      routeNames: [],
      type: 'bar',
      stale: false
    }
  },
  getParent() {
    throw new Error('not implemented')
  },
  setOptions() {}
}

export function fakeSceneProps<Name extends keyof AppParamList>(name: Name, params: AppParamList[Name]): EdgeSceneProps<Name> {
  return {
    navigation: fakeNavigation,
    route: {
      name: name,
      key: name + '-0',
      params
    } as any
  }
}
