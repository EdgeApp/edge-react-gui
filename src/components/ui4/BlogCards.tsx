import { BlogPost, BlogPostGeo } from 'edge-info-server'
import React from 'react'
import { ListRenderItem } from 'react-native'
import { useSafeAreaFrame } from 'react-native-safe-area-context'

import { useHandler } from '../../hooks/useHandler'
import { infoServerData } from '../../util/network'
import { useTheme } from '../services/ThemeContext'
import { BlogCard } from './BlogCard'
import { CarouselUi4 } from './CarouselUi4'

export interface Props {
  countryCode?: string
}

/**
 * Filters a list of blog posts based on the provided country code.
 *
 * @param cards - An array of `BlogPostGeo` objects representing the blog posts to filter.
 * @param countryCode - An optional string representing the country code to filter by. If `null`, an empty array is returned.
 * @returns An array of `BlogPostGeo` objects that match the provided country code.
 */
export const filterBlogCards = (cards: BlogPostGeo[], countryCode?: string) => {
  return cards.filter((blog: BlogPostGeo) => {
    const { countryCodes: includeCountryCodes = [], excludeCountryCodes = [] } = blog

    const isCountryInclude =
      includeCountryCodes.length === 0 ||
      (countryCode != null && includeCountryCodes.some(includeCountryCode => includeCountryCode.toUpperCase() === countryCode.toUpperCase()))
    const isCountryExclude =
      excludeCountryCodes.length > 0 &&
      (countryCode == null || excludeCountryCodes.some(excludeCountryCode => excludeCountryCode.toUpperCase() === countryCode.toUpperCase()))

    return isCountryInclude && !isCountryExclude
  })
}

/**
 * Renders a carousel of blog cards using the provided blog post data.
 *
 * @param props - The component props, including the country code to filter blog posts.
 * @returns A React element representing the blog cards carousel.
 */
export const BlogCards = (props: Props) => {
  const { countryCode } = props
  const { width } = useSafeAreaFrame()
  const theme = useTheme()

  const [blogCards, setBlogCards] = React.useState<BlogPost[]>([])

  const renderBlog: ListRenderItem<BlogPost> = useHandler(({ item }) => <BlogCard blogPost={item} />)

  // Check for BlogCards from info server:
  React.useEffect(() => {
    const filteredBlogPostsGeo = filterBlogCards(infoServerData.rollup?.blogPostsGeo ?? [], countryCode)

    setBlogCards([...(infoServerData.rollup?.blogPosts ?? []), ...filteredBlogPostsGeo])
  }, [countryCode])

  return <CarouselUi4 data={blogCards} renderItem={renderBlog} height={theme.rem(13)} width={width} />
}
