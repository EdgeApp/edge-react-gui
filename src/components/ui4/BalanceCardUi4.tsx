import { lt } from 'biggystring'
import * as React from 'react'
import { View } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { toggleAccountBalanceVisibility } from '../../actions/LocalSettingsActions'
import { getFiatSymbol } from '../../constants/WalletAndCurrencyConstants'
import { useHandler } from '../../hooks/useHandler'
import { useWatch } from '../../hooks/useWatch'
import { formatNumber } from '../../locales/intl'
import { lstrings } from '../../locales/strings'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { getTotalFiatAmountFromExchangeRates, removeIsoPrefix, zeroString } from '../../util/utils'
import { AnimatedNumber } from '../common/AnimatedNumber'
import { EdgeTouchableOpacity } from '../common/EdgeTouchableOpacity'
import { TransferModal } from '../modals/TransferModal'
import { Airship } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { ButtonsViewUi4 } from './ButtonsViewUi4'
import { CardUi4 } from './CardUi4'

// Numbers larger than this are likely to overflow the display width so don't
// use regular Text components which can auto shrink
const MAX_ANIMATED_AMOUNT = '10000000'
interface Props {
  navigation: NavigationBase
  onViewAssetsPress?: () => void
}

/**
 * Card that displays balance, deposit/send buttons, and a link to view assets
 */
export const BalanceCardUi4 = (props: Props) => {
  const { navigation, onViewAssetsPress } = props

  const dispatch = useDispatch()
  const theme = useTheme()
  const styles = getStyles(theme)

  const account = useSelector(state => state.core.account)
  const isBalanceVisible = useSelector(state => state.ui.settings.isAccountBalanceVisible)
  const defaultIsoFiat = useSelector(state => state.ui.settings.defaultIsoFiat)
  const fiatAmount = useSelector(state => getTotalFiatAmountFromExchangeRates(state, defaultIsoFiat))
  const exchangeRates = useSelector(state => state.exchangeRates)

  const activeWalletIds = useWatch(account, 'activeWalletIds')
  const currencyWallets = useWatch(account, 'currencyWallets')
  const currencyWalletErrors = useWatch(account, 'currencyWalletErrors')

  const exchangeRatesReady = React.useMemo(
    () =>
      activeWalletIds.every(walletId => {
        // Ignore wallets that have crashed:
        if (currencyWalletErrors[walletId]) return true

        // Both the wallet and its rate need to be loaded:
        const wallet = currencyWallets[walletId]
        return wallet != null && exchangeRates[`${wallet.currencyInfo.currencyCode}_${defaultIsoFiat}`] != null
      }),
    [activeWalletIds, currencyWalletErrors, currencyWallets, exchangeRates, defaultIsoFiat]
  )

  const fiatSymbol = defaultIsoFiat ? getFiatSymbol(defaultIsoFiat) : ''
  const fiatCurrencyCode = removeIsoPrefix(defaultIsoFiat)
  const formattedFiat = isBalanceVisible ? formatNumber(fiatAmount, { toFixed: 2 }) : lstrings.redacted_placeholder
  const handleToggleAccountBalanceVisibility = useHandler(() => {
    dispatch(toggleAccountBalanceVisibility())
  })

  const handleSend = useHandler(() => {
    Airship.show(bridge => <TransferModal depositOrSend="send" bridge={bridge} account={account} navigation={navigation} />).catch(() => {})
  })

  const handleDeposit = useHandler(() => {
    Airship.show(bridge => <TransferModal depositOrSend="deposit" bridge={bridge} account={account} navigation={navigation} />).catch(() => {})
  })

  const balanceString = fiatSymbol.length !== 1 ? `${formattedFiat} ${fiatCurrencyCode}` : `${fiatSymbol} ${formattedFiat} ${fiatCurrencyCode}`
  const animateNumber = lt(fiatAmount, MAX_ANIMATED_AMOUNT)

  return (
    <CardUi4>
      <EdgeTouchableOpacity style={styles.balanceContainer} onPress={handleToggleAccountBalanceVisibility}>
        <View style={styles.titleContainer}>
          <EdgeText style={theme.cardTextShadow}>{lstrings.fragment_wallets_balance_text}</EdgeText>
          <IonIcon name={isBalanceVisible ? 'eye-off-outline' : 'eye-outline'} style={styles.eyeIcon} color={theme.iconTappable} size={theme.rem(1)} />
        </View>
        <View style={styles.balanceTextContainer}>
          {!exchangeRatesReady && zeroString(fiatAmount) ? (
            <EdgeText style={styles.balanceText}>{lstrings.loading}</EdgeText>
          ) : animateNumber ? (
            <AnimatedNumber numberString={balanceString} textStyle={styles.balanceText} />
          ) : (
            <EdgeText style={styles.balanceTextNoAnim}>{balanceString}</EdgeText>
          )}
        </View>
      </EdgeTouchableOpacity>
      {onViewAssetsPress == null ? null : (
        <EdgeTouchableOpacity style={styles.rightButtonContainer} onPress={onViewAssetsPress}>
          <EdgeText style={styles.tappableText}>{lstrings.view_assets}</EdgeText>
        </EdgeTouchableOpacity>
      )}

      <ButtonsViewUi4
        layout="row"
        secondary={{
          onPress: handleDeposit,
          label: lstrings.loan_fragment_deposit
        }}
        secondary2={{
          onPress: handleSend,
          label: lstrings.fragment_send_subtitle
        }}
      />
    </CardUi4>
  )
}

const getStyles = cacheStyles((theme: Theme) => {
  const balanceTextContainer = {
    marginTop: theme.rem(0.25),
    marginBottom: theme.rem(0.5),
    height: theme.rem(2.25)
  }

  const balanceText = {
    fontSize: theme.rem(1.75),
    fontFamily: theme.fontFaceMedium,
    color: theme.primaryText,
    includeFontPadding: false,
    ...theme.cardTextShadow
  }

  return {
    balanceContainer: {
      margin: theme.rem(0.5)
    },
    titleContainer: {
      flexDirection: 'row',
      alignItems: 'center'
    },

    rightButtonContainer: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      position: 'absolute',
      top: 0,
      right: 0
    },
    tappableText: {
      fontSize: theme.rem(0.75),
      color: theme.iconTappable,
      margin: theme.rem(1),
      marginTop: theme.rem(1) + 3, // Fudge factor to align with the larger text on the left
      ...theme.cardTextShadow
    },

    // These two icons have different bounding boxes. Adjusted to match
    eyeIcon: {
      alignSelf: 'center',
      marginLeft: theme.rem(0.25),
      marginRight: theme.rem(0),
      ...theme.cardTextShadow
    },
    balanceText,
    balanceTextContainer,
    balanceTextNoAnim: {
      ...balanceTextContainer,
      ...balanceText
    },
    balanceBoxContainer: {
      height: theme.rem(3.25),
      marginTop: theme.rem(0.5)
    },
    balanceHeader: {
      fontSize: theme.rem(1),
      color: theme.secondaryText
    },
    showBalance: {
      fontSize: theme.rem(1.5),
      fontFamily: theme.fontFaceMedium
    }
  }
})
