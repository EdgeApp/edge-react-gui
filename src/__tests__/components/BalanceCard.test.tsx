import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import renderer from 'react-test-renderer'

import { BalanceCardUi4 } from '../../components/ui4/BalanceCardUi4'
import { FakeProviders } from '../../util/fake/FakeProviders'
import { fakeNavigation } from '../../util/fake/fakeSceneProps'

describe('BalanceCard', () => {
  it('should render with loading props', () => {
    const actual = renderer.create(
      <FakeProviders>
        <BalanceCardUi4 navigation={fakeNavigation} />
      </FakeProviders>
    )

    expect(actual).toMatchSnapshot()
  })
})
