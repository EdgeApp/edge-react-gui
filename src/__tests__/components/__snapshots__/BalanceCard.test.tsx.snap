// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalanceCard should render with loading props 1`] = `
<View
  style={
    [
      {
        "alignSelf": "stretch",
        "borderRadius": 16,
      },
      {
        "marginBottom": 11,
        "marginLeft": 11,
        "marginRight": 11,
        "marginTop": 11,
      },
      {
        "paddingBottom": 11,
        "paddingLeft": 11,
        "paddingRight": 11,
        "paddingTop": 11,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "backgroundColor": "rgba(255, 255, 255, .1)",
        "borderRadius": 16,
        "bottom": 0,
        "left": 0,
        "overflow": "hidden",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "margin": 11,
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        adjustsFontSizeToFit={true}
        minimumFontScale={0.65}
        numberOfLines={1}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "Quicksand-Regular",
              "fontSize": 22,
              "includeFontPadding": false,
            },
            {
              "textShadowColor": "rgba(0, 0, 0, 0.514)",
              "textShadowOffset": {
                "height": 0,
                "width": 0,
              },
              "textShadowRadius": 3,
            },
            null,
          ]
        }
      >
        Total Balance
      </Text>
      <Text
        allowFontScaling={false}
        style={
          [
            {
              "color": "#00f1a2",
              "fontSize": 22,
            },
            {
              "alignSelf": "center",
              "marginLeft": 6,
              "marginRight": 0,
              "textShadowColor": "rgba(0, 0, 0, 0.514)",
              "textShadowOffset": {
                "height": 0,
                "width": 0,
              },
              "textShadowRadius": 3,
            },
            {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        ïŠµ
      </Text>
    </View>
    <View
      style={
        {
          "height": 51,
          "marginBottom": 11,
          "marginTop": 6,
        }
      }
    >
      <Text
        adjustsFontSizeToFit={true}
        minimumFontScale={0.65}
        numberOfLines={1}
        style={
          [
            {
              "color": "#FFFFFF",
              "fontFamily": "Quicksand-Regular",
              "fontSize": 22,
              "includeFontPadding": false,
            },
            {
              "height": 51,
              "marginBottom": 11,
              "marginTop": 6,
              "textShadowColor": "rgba(0, 0, 0, 0.514)",
              "textShadowOffset": {
                "height": 0,
                "width": 0,
              },
              "textShadowRadius": 3,
            },
            null,
          ]
        }
      >
        Exchange Rates Loading...
      </Text>
    </View>
  </View>
  <View
    absolute={false}
    layout="row"
    style={
      {
        "flexDirection": "row-reverse",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "margin": 11,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "stretch",
          "flex": 1,
          "justifyContent": "center",
          "margin": -11,
          "opacity": 1,
          "padding": 11,
        }
      }
    >
      <BVLinearGradient
        colors={
          [
            872415231,
            872415231,
          ]
        }
        endPoint={
          {
            "x": 1,
            "y": 1,
          }
        }
        locations={null}
        startPoint={
          {
            "x": 0,
            "y": 0,
          }
        }
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 34,
              "flexDirection": "row",
              "flexGrow": 0,
              "flexShrink": 0,
              "justifyContent": "center",
            },
            {
              "alignSelf": "stretch",
            },
            {
              "height": 67,
              "paddingHorizontal": 34,
            },
            {
              "opacity": 1,
            },
          ]
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              [
                {
                  "color": "#FFFFFF",
                  "fontFamily": "Quicksand-Regular",
                  "fontSize": 22,
                },
                null,
              ],
              null,
            ]
          }
        >
          Deposit
        </Text>
      </BVLinearGradient>
    </View>
    <View
      style={
        [
          {
            "alignItems": undefined,
            "flex": undefined,
            "flexDirection": "column",
            "justifyContent": undefined,
            "marginBottom": 11,
            "marginLeft": 11,
            "marginRight": 11,
            "marginTop": 11,
          },
          undefined,
        ]
      }
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "stretch",
          "flex": 1,
          "justifyContent": "center",
          "margin": -11,
          "opacity": 1,
          "padding": 11,
        }
      }
    >
      <BVLinearGradient
        colors={
          [
            872415231,
            872415231,
          ]
        }
        endPoint={
          {
            "x": 1,
            "y": 1,
          }
        }
        locations={null}
        startPoint={
          {
            "x": 0,
            "y": 0,
          }
        }
        style={
          [
            {
              "alignItems": "center",
              "borderRadius": 34,
              "flexDirection": "row",
              "flexGrow": 0,
              "flexShrink": 0,
              "justifyContent": "center",
            },
            {
              "alignSelf": "stretch",
            },
            {
              "height": 67,
              "paddingHorizontal": 34,
            },
            {
              "opacity": 1,
            },
          ]
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              [
                {
                  "color": "#FFFFFF",
                  "fontFamily": "Quicksand-Regular",
                  "fontSize": 22,
                },
                null,
              ],
              null,
            ]
          }
        >
          Send
        </Text>
      </BVLinearGradient>
    </View>
  </View>
</View>
`;
